lib/tree.rb

```diff
@@ -16,40 +16,144 @@ def initialize
-  # Space Complexity: 
+  #Recursive helper method
+  def add_helper(current_node, key, value)
+    return TreeNode.new(key,value) if current_node.nil?
+    if key < current_node.key
+      current_node.left = add_helper(current_node.left, key, value)
+    else
+      current_node.right = add_helper(current_node.right, key, value)
+    end
+    return current_node
+  end
+
+  # Time Complexity: Olog(n) where n is the number of nodes it will traverse
+  # Space Complexity: Olog(n) where n is the number of function calls
   def find(key)
-    raise NotImplementedError
+    # raise NotImplementedError
+    return nil if @root == nil
+    find_help(@root, key)
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def find_help(current_node, key)
+    return nil if current_node.nil?
+    return current_node.value if current_node.key == key
+    # return nil if current_node.left == nil && current_node.right == nil
+
+    if key < current_node.key
+      current_node = current_node.left
+      find_help(current_node, key)
+    else
+      current_node = current_node.right
+      find_help(current_node,key)
+    end
+  end
+
+  # Time Complexity: O(n) where n is number of nodes
+  # Space Complexity: O(n) where n is the number of function calls
   def inorder
-    raise NotImplementedError
+    # raise NotImplementedError
+    return inorder_helper(@root, [])
   end
+  def inorder_helper(current_node,list)
+    return list if current_node.nil? #base case?
 
-  # Time Complexity: 
-  # Space Complexity: 
+    inorder_helper(current_node.left, list)
+    list << { key: current_node.key, value: current_node.value }
+    inorder_helper(current_node.right, list)
+    
+    return list
+  end
+
+  # Time Complexity: O(n) where n is number of nodes
+  # Space Complexity: O(n) where n is the number of function calls
   def preorder
-    raise NotImplementedError
+    # raise NotImplementedError
+    return preorder_helper(@root, [])
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def preorder_helper(current_node,list)
+    return list if current_node.nil?
+
+    list << { key: current_node.key, value: current_node.value }
+
+    preorder_helper(current_node.left, list)
+    preorder_helper(current_node.right, list)
+    
+    return list
+  end
+
+  # Time Complexity: O(n) where n is number of nodes
+  # Space Complexity: O(n) where n is the number of function calls
   def postorder
-    raise NotImplementedError
+    return postorder_helper(@root, [])
   end
+  def postorder_helper(current_node, list)
+    return list if current_node.nil?
 
-  # Time Complexity: 
-  # Space Complexity: 
+    postorder_helper(current_node.left, list)
+    postorder_helper(current_node.right, list)
+    list << { key: current_node.key, value: current_node.value }
+    
+    return list
+  end
+
+  # Time Complexity: O(n) where is equal to number of each single node in the tree
+  # Space Complexity: O(n) where n is the number of functino calls
   def height
```

Comment:
⚠️ This method isn't working.  It's always returning `height_helper(current_node.right, max, count+1)`